server/utils/utils.test.ts

```diff
@@ -382,209 +383,246 @@ describe('utils', () => {
-            offenderNo: 'ABC123',
-            reasonDescription: 'Victim',
-          },
-          {
-            agencyId: 'MDI',
-            assignedLivingUnitDescription: 'NMI-RECP',
-            nonAssociationName: 'Guy Incognito',
-            offenderNo: 'DEF321',
-            reasonDescription: 'Rival Gang',
-          },
-        ])
-        expect(res).toEqual([
-          [
-            { html: '<a class="govuk-link govuk-link--no-visited-state" href="/prisoner/ABC123">John Doe</a>' },
-            { text: 'ABC123' },
-            { text: 'NMI-RECP' },
-            { text: 'Victim' },
-          ],
-          [
-            { html: '<a class="govuk-link govuk-link--no-visited-state" href="/prisoner/DEF321">Guy Incognito</a>' },
-            { text: 'DEF321' },
-            { text: 'NMI-RECP' },
-            { text: 'Rival Gang' },
-          ],
-        ])
-      })
+    it('should cope with null location', () => {
+      expect(extractLocation(null, 'MDI')).toEqual(undefined)
+    })
+
+    it('should cope with null agencyId', () => {
+      expect(extractLocation('MDI-1-1-1', null)).toEqual(undefined)
+    })
+
+    it('returns undefined for locations that do not fit expected pattern', () => {
+      expect(extractLocation('UNEXPECTED', 'MDI')).toEqual(undefined)
+      expect(extractLocation('1-1-1', 'MDI')).toEqual(undefined)
+      expect(extractLocation('LEI-1-1-1', 'MDI')).toEqual(undefined)
+    })
+
+    it('extracts locations', () => {
+      expect(extractLocation('MDI-1-1-1', 'MDI')).toEqual('1-1-1')
+    })
+
+    it('should extract RECP,CSWAP,COURT', () => {
+      expect(extractLocation('MDI-RECP', 'MDI')).toEqual('Reception')
+      expect(extractLocation('MDI-CSWAP', 'MDI')).toEqual('No cell allocated')
+      expect(extractLocation('MDI-COURT', 'MDI')).toEqual('Court')
+    })
+  })
+
+  describe('toNonAssociationRows()', () => {
+    it('map non-associations to rows', () => {
+      const res = toNonAssociationRows([
+        {
+          agencyId: 'MDI',
+          assignedLivingUnitDescription: 'NMI-RECP',
+          nonAssociationName: 'John Doe',
+          offenderNo: 'ABC123',
+          reasonDescription: 'Victim',
+        },
+        {
+          agencyId: 'MDI',
+          assignedLivingUnitDescription: 'NMI-RECP',
+          nonAssociationName: 'Guy Incognito',
+          offenderNo: 'DEF321',
+          reasonDescription: 'Rival Gang',
+        },
+      ])
+      expect(res).toEqual([
+        [
+          { html: '<a class="govuk-link govuk-link--no-visited-state" href="/prisoner/ABC123">John Doe</a>' },
+          { text: 'ABC123' },
+          { text: 'NMI-RECP' },
+          { text: 'Victim' },
+        ],
+        [
+          { html: '<a class="govuk-link govuk-link--no-visited-state" href="/prisoner/DEF321">Guy Incognito</a>' },
+          { text: 'DEF321' },
+          { text: 'NMI-RECP' },
+          { text: 'Rival Gang' },
+        ],
+      ])
     })
   })
```

Comment:
There were some tests that were embedded at the wrong level (grouped with the `getError` tests) so I've moved them out.